File: sort/quick_sort/javaScript/quick_sort.js
//Like Merge Sort, QuickSort is a Divide and Conquer algorithm. It picks an element as pivot and partitions the given array around the picked pivot. There are many different versions of quickSort that pick pivot in different ways.
//
//    Always pick first element as pivot.
//    Always pick last element as pivot (implemented below)
//    Pick a random element as pivot.
//    Pick median as pivot.


function quickSort(arr, left, right){
   var pivot;
   var partitionIndex;

  if(left < right){
    pivot = right;
    partitionIndex = partition(arr, pivot, left, right);
    quickSort(arr, left, partitionIndex - 1);
    quickSort(arr, partitionIndex + 1, right);
  }
  return arr;
}

function partition(arr, pivot, left, right){
   var pivotValue = arr[pivot];
   var partitionIndex = left;

   for(var i = left; i < right; i++){
    if(arr[i] < pivotValue){
      swap(arr, i, partitionIndex);
      partitionIndex++;
    }
  }
  swap(arr, right, partitionIndex);
  return partitionIndex;
}

function swap(arr, i, j){
   [arr[i], arr[j]] = [arr[j], arr[i]];
}

File: sort/quick_sort/c#/quicksort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Quicksort
{
	class Program
	{
		static void Main(string[] args)
		{
			int[] arr = { 123, 12, 912, 16, 83, 1, -12, 1203};

			Console.WriteLine(string.Join(",", arr));

			Quicksort(arr, 0, arr.Length - 1);

			Console.WriteLine(string.Join(",", arr));
		}

		public static void Quicksort(int[] arr, int left, int right)
		{
			int i = left, j = right;
			int pivot = arr[(left + right) / 2];

			while (i <= j)
			{
				while (arr[i].CompareTo(pivot) < 0)
				{
					i++;
				}

				while (arr[j].CompareTo(pivot) > 0)
				{
					j--;
				}

				if (i <= j)
				{
					int tmp = arr[i];
					arr[i] = arr[j];
					arr[j] = tmp;

					i++;
					j--;
				}
			}

			if (left < j)
			{
				Quicksort(arr, left, j);
			}
			if (i < right)
			{
				Quicksort(arr, i, right);
			}
		}
	}
}
